File: Algorithms/0554_Brick_Wall/Java/BrickWallSolution2.java

// Space: O(n)
// Time: O(n)

public class Solution {
    public int leastBricks(List<List<Integer>> wall) {
        int height = wall.size();
        if (height == 1) {
            return wall.get(0).size() == 1 ? 1 : 0;
        }

        Map<Integer, Integer> cache = new HashMap<>();
        int maxEdge = 0;

        for (List<Integer> row : wall) {
            for (int i = 1; i < row.size(); i++) {
                int edge = row.subList(0, i).stream().reduce(Integer::sum).get();
                if (cache.containsKey(edge)) {
                    cache.put(edge, cache.get(edge) + 1);
                } else {
                    cache.put(edge, 1);
                }
                maxEdge = cache.get(edge) > maxEdge ? cache.get(edge) : maxEdge;
            }
        }
        return height - maxEdge;
    }
}

File: Algorithms/0554_Brick_Wall/Python/Brick_Wall_Solution_2.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def leastBricks(self, wall):
        height = len(wall)
        if height == 1:
            return 0 if len(wall[0]) > 1 else 1
        cache, max_edge = collections.defaultdict(int), 0

        for row in wall:
            for i in range(1, len(row)):
                edge = sum(row[:i])
                cache[edge] += 1
                max_edge = max(max_edge, cache[edge])

        return height - max_edge




